pkg/crd/testdata/oneof/types.go

```diff
@@ -0,0 +1,86 @@
+/*
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// +groupName=testdata.kubebuilder.io
+// +versionName=v1beta1
+package oneof
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+// +kubebuilder:resource:singular=oneof
+
+// OneofSpec is the spec for the oneofs API.
+type OneofSpec struct {
```

Comment:
So further to my previous comment, lets see if we can add an example where it would be `// +kubebuilder:validation:AtMostOneOf=foo.bar;bar`, I suspect this would still work

Reply:
Moreover, we have validations to correctly mark oneof annotated fields as Optional (not Required) in structToSchema. Allowing for nested field access here would imply we can't assume the field names are the fields belonging to the `Type` being evaluated in `structToSchema`, as they would be nested fields. 

I prefer to not allow nested oneof constaints right now, as almost all use cases pertain to top-level fields. If there is an ask for it, we could introduce it in the future.

Reply:
I added a validation check to prevent nested field references. Supporting nested fields is a lot more complex and we don't require it for a majority of use cases, and can be added as a future enhancement.